src/chainstate/stacks/index/marf.rs

```diff
@@ -0,0 +1,2228 @@
+use std::fs;
+
+use sha2::Digest;
+
+use chainstate::burn::BlockHeaderHash;
+
+use chainstate::stacks::index::bits::{
+    get_node_hash,
+    get_node_hash_bytes
+};
+
+use chainstate::stacks::index::node::{
+    TrieNodeID,
+    TrieNodeType,
+    TrieNode,
+    TrieNode4,
+    TrieNode16,
+    TrieNode48,
+    TrieNode256,
+    TrieLeaf,
+    TriePtr,
+    TRIEPTR_SIZE,
+    TrieCursor,
+    TriePath,
+    is_backptr,
+    set_backptr,
+    clear_backptr,
+};
+
+use chainstate::stacks::index::storage::{
+    read_all,
+    write_all,
+    fseek,
+    fseek_end,
+    TrieStorage,
+    TrieFileStorage
+};
+
+use chainstate::stacks::index::{
+    TrieHash,
+    TRIEHASH_ENCODED_SIZE,
+    fast_extend_from_slice,
+    MARFValue
+};
+
+use chainstate::stacks::index::trie::{
+    Trie,
+};
+
+use chainstate::stacks::index::Error as Error;
+
+use util::log;
+
+
+/// Merklized Adaptive-Radix Forest -- a collection of Merklized Adaptive-Radix Tries.
+pub struct MARF<S>
+where
+    S: TrieStorage + Seek
+{
+    storage: S,
+    open_chain_tip: Option<BlockHeaderHash>
+}
+
+impl<S> MARF<S>
+where
+    S: TrieStorage + Seek
+{
+
+    // helper method for walking a node's backpr
+    fn walk_backptr(s: &mut S, start_node: &TrieNodeType, chr: u8, c: &mut TrieCursor) -> Result<(TrieNodeType, TrieHash, TriePtr, u32), Error> {
+        let ptr_opt = match start_node {
+            TrieNodeType::Node4(ref data) => data.walk(chr),
+            TrieNodeType::Node16(ref data) => data.walk(chr),
+            TrieNodeType::Node48(ref data) => data.walk(chr),
+            TrieNodeType::Node256(ref data) => data.walk(chr),
+            _ => {
+                panic!("Did not get an intermediate node");
+            }
+        };
+        match ptr_opt {
+            None => {
+                // this node never had a child for this chr
+                trace!("Failed to walk to '{}' from {:?}", chr, start_node);
+                Err(Error::BackptrNotFoundError)
+            },
+            Some(ptr) => {
+                trace!("Walk backptrs for {:?} to {:?} from {:?}", c, &ptr, &start_node);
+                
+                // this node had a child for this chr at one point
+                let (node, node_hash, node_ptr) = match start_node {
+                    TrieNodeType::Node4(_) => Trie::walk_backptr(s, &ptr, c)?,
+                    TrieNodeType::Node16(_) => Trie::walk_backptr(s, &ptr, c)?,
+                    TrieNodeType::Node48(_) => Trie::walk_backptr(s, &ptr, c)?,
+                    TrieNodeType::Node256(_) => Trie::walk_backptr(s, &ptr, c)?,
+                    _ => {
+                        unreachable!();
+                    }
+                };
+
+                Ok((node, node_hash, node_ptr, ptr.back_block))
+            }
+        }
+    }
+   
+    fn node_copy_update(_s: &mut S, node: &mut TrieNodeType, node_dist: u32) -> Result<TrieHash, Error> {
+        fn node_copy_update_ptrs(ptrs: &mut [TriePtr], node_dist: u32) -> () {
```

Comment:
Why define this function in the local scope of another function?

Reply:
This is the only function that uses it, and it's only useful for this purpose.  I can move it out as a private function in the `MARF` impl if it would improve code quality.

Reply:
This should be addressed now.